### PR TITLE
Sync employees to biometric devices

### PR DESCRIPTION
**Pull request by KaviyaPeriyasamy**
This PR helps to achieve user management

Screenshot from 2021-01-12 11-00-53

employee

settings

### CODE REVIEW

File: essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py [hunk 1]
```diff
@@ -96,4 +127,121 @@ def calculate_total_shift(shift_doc, logs):
...
+			settings = frappe.get_single('Essdee Attendance Settings')
+			for device in settings.device_details:
+				device_doc = frappe.get_doc('Essdee Biometric Device', device.device_id)
+				employee_record = frappe.get_all('Employee', filters = [["Work Location", "sd_location", "=", device_doc.location]], fields = ['attendance_device_id','employee_name'])
+				conn = sync_device(ip = device_doc.ip)
+				if conn:
+					for record in employee_record:
+						conn.set_user(uid= record['attendance_device_id'], name= record['employee_name'], user_id= record['attendance_device_id'])
+					sync_fingerprint(conn)
+					conn.disconnect()
```
Comment (karthikeyan5) [inline essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py:179]: refactor as discussed

File: essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py [hunk 2]
```diff
@@ -96,4 +127,121 @@ def calculate_total_shift(shift_doc, logs):
...
+			for device in settings.device_details:
+				device_doc = frappe.get_doc('Essdee Biometric Device', device.device_id)
+				if device_doc.location == row['sd_location']:
+					conn = sync_device(ip = device_doc.ip)
+					if conn:
+						conn.enroll_user(id)
+						conn.disconnect()
+		msgprint(_('Successfully enrolled'))
+	except Exception as e:
+		raise e
```
Comment (karthikeyan5) [inline essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py:219]: also, give provision for fingerprint enrollment.

Comment (karthikeyan5) [reply essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py:219]: as discussed, I will check with the device and let you know.

File: essdee_attendance/essdee_attendance/doctype/essdee_biometric_device/essdee_biometric_device.json [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+   "fieldname": "sync_now",
+   "fieldtype": "Button",
+   "label": "Sync Now"
+  },
+  {
+   "fieldname": "combined_shift_type",
+   "fieldtype": "Table MultiSelect",
+   "label": "Combined Shift Type",
+   "options": "Combined Shift Type",
+   "reqd": 1
```
Comment (karthikeyan5) [inline essdee_attendance/essdee_attendance/doctype/essdee_biometric_device/essdee_biometric_device.json:50]: note to self: check if last sync works as expected.

**Review by karthikeyan5 — CHANGES_REQUESTED** [review]

File: essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py (hunk 1 above)
Comment (KaviyaPeriyasamy) [reply essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py]: Refactored here https://github.com/aerele/essdee_attendance/pull/2/commits/a994056abba2e42ca32e9d768b138c847c8c4d4e

File: essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py (hunk 2 above)
Comment (KaviyaPeriyasamy) [reply essdee_attendance/essdee_attendance/doctype/essdee_attendance_settings/essdee_attendance_settings.py]: Checked fingerprint enrollment functionality in the device

File: essdee_attendance/essdee_attendance/doctype/essdee_biometric_device/essdee_biometric_device.json (hunk 3 above)
Comment (KaviyaPeriyasamy) [reply essdee_attendance/essdee_attendance/doctype/essdee_biometric_device/essdee_biometric_device.json:51]: The shift-type last sync works as expected

**Pull request merged**